src/base64_ops.rs

```diff
@@ -0,0 +1,96 @@
+use crate::allocator::{Allocator, Atom, NodePtr};
+use crate::cost::check_cost;
+use crate::cost::Cost;
+use crate::err_utils::err;
+use crate::op_utils::{atom_len, get_args, MALLOC_COST_PER_BYTE};
+use crate::reduction::{EvalErr, Reduction, Response};
+use crate::SExp;
+use base64ct::{Base64UrlUnpadded, Encoder, Encoding};
+
+const BASE64_ENCODE_BASE_COST: Cost = 40;
+const BASE64_COST_PER_ARG: Cost = 130;
+const BASE64_DECODE_BASE_COST: Cost = 400;
+const BASE64_COST_PER_BYTE: Cost = 3;
+
+const fn encoded_len(n: usize) -> Option<usize> {
+    let Some(q) = n.checked_mul(4) else {
+        return None;
+    };
+    Some((q / 3) + (q % 3 != 0) as usize)
+}
+
+pub fn op_base64url_encode(a: &mut Allocator, input: NodePtr, max_cost: Cost) -> Response {
+    let mut cost = BASE64_ENCODE_BASE_COST;
+    let mut input_size: usize = 0;
+
+    let mut arg_list = input;
+    while let Some((arg, rest)) = a.next(arg_list) {
+        arg_list = rest;
+        cost += BASE64_COST_PER_ARG;
+        let len = match a.sexp(arg) {
+            SExp::Pair(_, _) => return err(arg, "base64url on list"),
+            SExp::Atom => a.atom_len(arg),
+        };
+
+        input_size += len;
+        cost += len as Cost * BASE64_COST_PER_BYTE;
+        check_cost(a, cost, max_cost)?;
+    }
+
+    if input_size == 0 {
+        return Ok(Reduction(cost, a.nil()));
+    }
+
+    let output_size =
+        encoded_len(input_size).ok_or(EvalErr(input, "base64 invalid input length".to_string()))?;
+
+    cost += output_size as Cost * MALLOC_COST_PER_BYTE;
+    check_cost(a, cost, max_cost)?;
+
+    let mut output: Vec<u8> = vec![0; output_size];
+    let mut enc = Encoder::<Base64UrlUnpadded>::new(&mut output[..])
+        .map_err(|_| EvalErr(input, "base64 (internal error)".to_string()))?;
```

Comment:
I'm wondering if it makes sense to `impl From<base64ct::Error>` so you could use the `?` operator directly on all of these errors (ane change the error message later in one place)

Reply:
in the long run, I think we need to clean up the error type situation in `clvm_rs` to have an error type that can wrap other error categories. like in `chia_rs`. It seems a bit wrong to add an implicit conversion between these errors that assume it's an internal error. In the decode case, it's not. In an ideal world we would preserve the base64 error code in that case.

Reply:
implementing an implicit conversion won't work because we also need the `NodePtr` argument for `EvalErr`

Reply:
Ah I missed that, good point